File: 2023/src/day2.go
// Day 2 - Now in Go!
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"strconv"
	"strings"
)

type Game struct {
	id     int
	rounds []Round
}

func (g Game) is_possible(red int, green int, blue int) bool {
	for _, round := range g.rounds {
		if !round.is_possible(red, green, blue) {
			return false
		}
	}
	return true
}

func (g Game) min_needed() []int {
	min_red, min_green, min_blue := 0, 0, 0

	for _, round := range g.rounds {
		if round.red > min_red {
			min_red = round.red
		}

		if round.green > min_green {
			min_green = round.green
		}

		if round.blue > min_blue {
			min_blue = round.blue
		}
	}

	return []int{min_red, min_green, min_blue}
}

type Round struct {
	red   int
	green int
	blue  int
}

func (r Round) is_possible(red int, green int, blue int) bool {
	return (r.red <= red) && (r.green <= green) && (r.blue <= blue)
}

type Day2 struct {
	games []Game
}

func (d Day2) part_1() int {
	min_red, min_green, min_blue := 12, 13, 14

	id_total := 0
	for _, game := range d.games {
		if game.is_possible(min_red, min_green, min_blue) {
			id_total += game.id
		}
	}

	return id_total
}

func (d Day2) part_2() int {
	var minimums [][]int
	sum_of_powers := 0

	for _, game := range d.games {
		minimums = append(minimums, game.min_needed())
	}

	for _, mins := range minimums {
		sum_of_powers += mins[0] * mins[1] * mins[2]
	}

	return sum_of_powers
}

func create_round(round_input_str string) Round {
	tokens := strings.Split(round_input_str, ",")
	red, green, blue := 0, 0, 0

	for _, token := range tokens {
		pair := strings.Split(strings.TrimSpace(token), " ")
		switch pair[1] {
		case "red":
			red, _ = strconv.Atoi(pair[0])
		case "green":
			green, _ = strconv.Atoi(pair[0])
		case "blue":
			blue, _ = strconv.Atoi(pair[0])
		}
	}

	return Round{red, green, blue}
}

func create_game(game_line string) Game {
	tokens := strings.Split(game_line, ":")
	id, _ := strconv.Atoi(strings.Split(strings.TrimSpace(tokens[0]), " ")[1])
	var rounds []Round

	for _, round_str := range strings.Split(tokens[1], ";") {
		rounds = append(rounds, create_round(round_str))
	}

	return Game{id: id, rounds: rounds}
}

func create_day_2(lines []string) Day2 {
	var games []Game
	for _, line := range lines {
		games = append(games, create_game(line))
	}

	return Day2{games}
}

func main() {
	file, err := os.Open("./input/day2.txt")

	if err != nil {
		log.Fatal(err)
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	var lines []string
	for scanner.Scan() {
		lines = append(lines, scanner.Text())
	}

	day_2 := create_day_2(lines)
	fmt.Println("Part 1 total:", day_2.part_1())
	fmt.Println("Part 2 total:", day_2.part_2())
}

File: 2023/src/day2.rb
class Game
  attr_reader :id
  attr_reader :rounds

  def initialize(game_line)
    tokens = game_line.split(':')
    @id = tokens[0].sub('Game ', '').to_i
    @rounds = tokens[1].split(';').map {|token| Round.new(token.strip)}
  end

  def possible?(red, green, blue)
    @rounds.each do |round|
      return false if not round.possible?(red, green, blue)
    end

    true
  end

  def min_needed
    min_red = 0
    min_green = 0
    min_blue = 0

    @rounds.each do |round|
      min_red = round.red if round.red > min_red
      min_green = round.green if round.green > min_green
      min_blue = round.blue if round.blue > min_blue
    end

    [min_red, min_green, min_blue]
  end
end

class Round
  attr_reader :red
  attr_reader :green
  attr_reader :blue

  def initialize(round_input_str)
    @red = 0
    @green = 0
    @blue = 0

    round_input_str.split(',').each do |token|
      pair = token.strip.split(' ')
      case pair[1]
      when 'red'
        @red = pair[0].to_i
      when 'green'
        @green = pair[0].to_i
      when 'blue'
        @blue = pair[0].to_i
      end
    end
  end

  def possible?(red, green, blue)
    (@red <= red) && (@green <= green) && (@blue <= blue)
  end
end


class Day2
  def initialize(input_lines)
    @games = input_lines.map {|line| Game.new(line)}
  end

  def part_1
    max_red = 12
    max_green = 13
    max_blue = 14

    @games.sum {|game| game.possible?(max_red, max_green, max_blue) ? game.id : 0}
  end

  def part_2
    minimums = @games.map {|game| game.min_needed}
    minimums.sum {|arr| arr[0] * arr[1] * arr[2]}
  end
end

day_2 = Day2.new(IO.readlines('./input/day2.txt'))
puts "Part 1 total: #{day_2.part_1}"
puts "Part 2 total: #{day_2.part_2}"
